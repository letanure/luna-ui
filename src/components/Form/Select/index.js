import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import FontAwesomeIcon from '@fortawesome/react-fontawesome';

/* eslint-disable no-extra-parens */
const Select = styled.select`
  order:2;
  width: 100%;
  border: none;
  height: 40px;
  outline: none;
  padding: 10px;
  font-size: 14px;
  box-shadow: none;
  appearance: none;
  -ms-appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
  transition: all .25s linear;
  background-color: ${(props) => props.theme.form.inputText.background};
  ${(props) =>
    (`
      border-radius: 2px;
      border: 1px solid ${props.theme.form.inputText.border};
    `)
}
  ${(props) => (
    props.icon.length > 0 ? `
      padding-left: 40px;
    ` : ''
  )}
  ::-webkit-input-placeholder { /* Chrome/Opera/Safari */
    color: ${(props) => props.theme.form.inputText.placeholder};
  }
  ::-moz-placeholder { /* Firefox 19+ */
    color: ${(props) => props.theme.form.inputText.placeholder};
  }
  :-ms-input-placeholder { /* IE 10+ */
    color: ${(props) => props.theme.form.inputText.placeholder};
  }
  :-moz-placeholder { /* Firefox 18- */
    color: ${(props) => props.theme.form.inputText.placeholder};
  }
  &:focus ~ span {
    color: ${(props) => props.theme.form.inputText.types[props.kind]};
  }
  &:focus {
    border-color: ${(props) => props.theme.form.inputText.types[props.kind]};
  }
  &:disabled ~ span {
    cursor: not-allowed;
    color: ${(props) => props.theme.form.inputText.disabledText};
  }
  &:disabled {
    cursor: not-allowed;
    color: ${(props) => props.theme.form.inputText.disabledText};
    background-color: ${(props) => props.theme.form.inputText.disabled};
  }
`;

const InputWrapper = styled.div`
  position: relative;
  width: 100%;
`;

const SelectIcon = styled.span`
  font-size: 14px;
  transition: all .25s linear;
  color: ${(props) => props.theme.form.inputText.placeholder};
  position: absolute;
  left: 15px;
  top: 13px;
`;

const ArrowIcon = styled.span`
  font-size: 14px;
  position: absolute;
  right: 15px;
  top: 13px;
`;

/**
 * Select component.
 * E.g.:
 * ```html
 *  <Form.Select icon="seedling" kind="default">
 *    <option>Angeratum</option>
 *    <option>Marigold</option>
 *  </Form.Select>
 * ```
 */
class SelectComponent extends React.PureComponent {
  render() {
    const { kind, icon, config, children, className } = this.props;

    if (icon.length > 0) {
      return (
        <InputWrapper>
          <Select kind={kind} icon={icon} {...config}>
            {children}
          </Select>
          <ArrowIcon>
            <FontAwesomeIcon icon="caret-down"/>
          </ArrowIcon>
          <SelectIcon>
            <FontAwesomeIcon icon={icon}/>
          </SelectIcon>
        </InputWrapper>
      );
    }

    return (
      <InputWrapper className={className}>
        <Select kind={kind} icon={icon} {...config}>
          {children}
        </Select>
        <ArrowIcon>
          <FontAwesomeIcon icon="caret-down"/>
        </ArrowIcon>
      </InputWrapper>
    );
  }
}

SelectComponent.propTypes = {
  children: PropTypes.node,
  icon: PropTypes.string,
  kind: PropTypes.string,
  config: PropTypes.object,
  className: PropTypes.string
};

SelectComponent.defaultProps = {
  children: null,
  icon: '',
  kind: 'primary',
  config: {},
  className: ''
};

export default SelectComponent;
