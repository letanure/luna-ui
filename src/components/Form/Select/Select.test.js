import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../../config/defaultTheme';
import ThemeProvider from '../../Theme/index';

import Select from './index';

describe('<Select />', () => {
  let selectList;
  let onClick;

  beforeEach(() => {
    onClick = sinon.spy();

    selectList =
      <ThemeProvider theme={theme}>
        <Select icon="seedling" kind="default" config={{ defaultValue: 2 }}>
          <option value="1">Angeratum</option>
          <option value="2">Marigold</option>
        </Select>
        <Select kind="default" config={{ onClick }}>
          <option>Dog</option>
          <option>Cat</option>
          <option>Rat</option>
        </Select>
      </ThemeProvider>;
  });

  test('Check options length', () => {
    const wrapper = mount(selectList);
    expect(wrapper.find('select').at(0)
      .find('option')).toHaveLength(2);
  });

  test('Check options deafult value', () => {
    const wrapper = mount(selectList);
    expect(wrapper.find('select').at(0)
      .getDOMNode().value).toBe('2');
  });

  test('Call onClick', () => {
    const wrapper = mount(selectList);
    const select = () => wrapper.find('select').at(1);
    select().simulate('click');
    expect(onClick).toHaveProperty('callCount', 1);
  });

  test('toMatchSnapshot', () => {
    const tree = renderer
      .create(selectList)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
