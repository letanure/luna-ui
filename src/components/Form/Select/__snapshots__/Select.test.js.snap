// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> toMatchSnapshot 1`] = `
<div>
  <div
    className="sc-bwzfXH bVXEpg"
  >
    <select
      className="sc-bdVaJa gThJEg"
      defaultValue={2}
      icon="seedling"
      kind="default"
    >
      <option
        value="1"
      >
        Angeratum
      </option>
      <option
        value="2"
      >
        Marigold
      </option>
    </select>
    <span
      className="sc-bxivhb fozrMO"
    />
    <span
      className="sc-htpNat eeDWLn"
    />
  </div>
  <div
    className="sc-bwzfXH bVXEpg"
  >
    <select
      className="sc-bdVaJa hcUPcu"
      icon=""
      kind="default"
      onClick={[Function]}
    >
      <option>
        Dog
      </option>
      <option>
        Cat
      </option>
      <option>
        Rat
      </option>
    </select>
    <span
      className="sc-bxivhb fozrMO"
    />
  </div>
</div>
`;
